openfisca_core/tracers.py

```diff
@@ -0,0 +1,135 @@
+              }
+
+    """
+    def __init__(self):
+        log.warn("The tracer is a feature that is still currently under experimentation. You are very welcome to use it and send us precious feedback, but keep in mind that the way it is used and the results it gives might change without any major version bump.")
+        self.requested_calculations = set()
+        self.stack = []
+        self.trace = {}
+        self._computation_log = []
+
+    def clone(self):
+        new = Tracer()
+        new.requested_calculations = copy.copy(self.requested_calculations)
+        new.stack = copy.copy(self.stack)
+        new.trace = copy.deepcopy(self.trace)
+        new._computation_log = copy.copy(self._computation_log)
+        return new
+
+    @staticmethod
+    def _get_key(variable_name, period, **parameters):
+        if parameters.get('extra_params'):
+            return u"{}<{}><{}>".format(variable_name, period, '><'.join(map(str, parameters['extra_params']))).encode('utf-8')
+        return u"{}<{}>".format(variable_name, period).encode('utf-8')
+
+    def record_calculation_start(self, variable_name, period, **parameters):
+        """
+            Record that OpenFisca started computing a variable.
+
+            :param str variable_name: Name of the variable starting to be computed
+            :param Period period: Period for which the variable is being computed
+            :param list parameters: Parameter with which the variable is being computed
+        """
+        key = self._get_key(variable_name, period, **parameters)
+
+        if self.stack:  # The variable is a dependency of another variable
+            parent = self.stack[-1]
+            self.trace[parent]['dependencies'].append(key)
+        else:  # The variable has been requested by the client
+            self.requested_calculations.add(key)
+
+        if not self.trace.get(key):
+            self.trace[key] = {'dependencies': []}
+        self.stack.append(key)
+        self._computation_log.append((key, len(self.stack)))
+
+    def record_calculation_end(self, variable_name, period, result, **parameters):
+        """
+            Record that OpenFisca finished computing a variable.
+
+            :param str variable_name: Name of the variable starting to be computed
+            :param Period period: Period for which the variable is being computed
+            :param list parameters: Parameter with which the variable is being computed
+        """
+        key = self._get_key(variable_name, period, **parameters)
+        expected_key = self.stack.pop()
+
+        if not key == expected_key:
+            raise ValueError(
+                u"Something went wrong with the simulation tracer: result of '{0}' was expected, got results for '{1}' instead. This does not make sense as the last variable we started computing was '{0}'."
+                .format(expected_key, key).encode('utf-8')
+                )
+        self.trace[key]['value'] = result.tolist()
```

Comment:
Would it be hepful to indicate that `.tolist()` transforms nparray into python lists ?

Reply:
Ok